modules/reporting/app/workers/cost_query/pdf/timesheet_generator.rb

```diff
@@ -0,0 +1,385 @@
+class CostQuery::PDF::TimesheetGenerator
+  include WorkPackage::PDFExport::Common::Common
+  include WorkPackage::PDFExport::Export::Cover
+  include WorkPackage::PDFExport::Common::Logo
+  include WorkPackage::PDFExport::Export::Page
+  include WorkPackage::PDFExport::Export::Style
+
+  H1_FONT_SIZE = 26
+  H1_MARGIN_BOTTOM = 2
+  HR_MARGIN_BOTTOM = 16
+  TABLE_CELL_FONT_SIZE = 10
+  TABLE_CELL_BORDER_COLOR = "BBBBBB".freeze
+  TABLE_CELL_PADDING = 4
+  COMMENT_FONT_COLOR = "636C76".freeze
+  H2_FONT_SIZE = 20
+  H2_MARGIN_BOTTOM = 10
+  COLUMN_DATE_WIDTH = 66
+  COLUMN_ACTIVITY_WIDTH = 100
+  COLUMN_HOURS_WIDTH = 60
+  COLUMN_TIME_WIDTH = 110
+  COLUMN_WP_WIDTH = 190
+
+  attr_accessor :pdf
+
+  def initialize(query, project)
+    @query = query
+    @project = project
+    @total_page_nr = nil
+    @page_count = 1
+    setup_page!
+  end
+
+  def heading
+    query.name || I18n.t(:"export.timesheet.timesheet")
+  end
+
+  def footer_title
+    heading
+  end
+
+  def cover_page_title
+    "OpenProject"
+  end
+
+  def cover_page_heading
+    heading
+  end
+
+  def cover_page_dates
+    dates_range = all_entries.group_by(&:spent_on).sort
+    start_date = dates_range.first&.first
+    end_date = dates_range.last&.first
```

Comment:
```suggestion
    start_date, end_date = all_entries.map(&:spent_on).minmax
```

Just to show off some nice ruby functions :)

Reply:
That is nicer indeed. ^^